components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java

```diff
@@ -78,19 +107,104 @@ public Response apisCopyApiPost(String newVersion,String apiId)
+            }
+            org.wso2.carbon.apimgt.api.model.API api = apiProvider.getAPI(apiIdentifier);
+            if (api != null) {
+
+                apiToReturn = MappingUtil.fromAPItoDTO(api);
+
+            } else {
+                //log the error
+                return Response.status(Response.Status.NOT_FOUND).entity("Cannot find the requested API- " + apiName +
+                        "-" + version).type(MediaType.APPLICATION_JSON).build();
+            }
+        } catch (APIManagementException e) {
+            //500
+            e.printStackTrace();
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity("Internal server error")
+                    .type(MediaType.APPLICATION_JSON).build();
+        } finally {
+            if (isTenantFlowStarted) {
+                PrivilegedCarbonContext.endTenantFlow();
+            }
+        }
+
+        return Response.ok().entity(apiToReturn).build();
     }
     @Override
     public Response apisApiIdPut(String apiId,API body,String contentType,String ifMatch,String ifUnmodifiedSince)
     throws NotFoundException {
-        // do some magic!
+        boolean isTenantFlowStarted = false;
+        try {
+            org.wso2.carbon.apimgt.api.model.API apiToAdd = MappingUtil.fromDTOtoAPI(body);
+
+            String tenantDomain = CarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+            if(tenantDomain != null && !MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)){
+                isTenantFlowStarted = true;
+                PrivilegedCarbonContext.startTenantFlow();
+            }
+            provider.updateAPI(apiToAdd);
+        } catch (APIManagementException e) {
+            //500
+            e.printStackTrace();
+        } catch (FaultGatewaysException e) {
+            e.printStackTrace();
+        } finally {
+            if (isTenantFlowStarted) {
+                PrivilegedCarbonContext.endTenantFlow();
+            }
+        }
+
         return Response.ok().entity(new ApiResponseMessage(ApiResponseMessage.OK, "magic!")).build();
     }
     @Override
     public Response apisApiIdDelete(String apiId,String ifMatch,String ifUnmodifiedSince)
     throws NotFoundException {
-        // do some magic!
+        String[] apiIdDetails = apiId.split("-");
+        String apiName = apiIdDetails[0];
+        String version = apiIdDetails[1];
+        String providerName = apiIdDetails[2];
+        String providerNameEmailReplaced = APIUtil.replaceEmailDomain(providerName);
+        APIIdentifier apiIdentifier = new APIIdentifier(providerNameEmailReplaced, apiName, version);
+        boolean isTenantFlowStarted = false;
+        try{
+            String tenantDomain = CarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+            if(tenantDomain != null && !MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
```

Comment:
Please note that if the tenant domain becomes null, !MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain) becomes true. Therefore it will start a tenant flow for null.
In order to guard this, tenantDomain != null is used.